iroh/src/magicsock/relay_actor.rs

```diff
@@ -104,59 +110,37 @@ impl ConnectedRelayActor {
+
+        let mut relay_send_fut = Either::Left(future::pending());
 
         loop {
             // If a read error occurred on the connection it might have been lost.  But we
             // need this connection to stay alive so we can receive more messages sent by
             // peers via the relay even if we don't start sending again first.
             if !self.relay_client.is_connected().await? {
                 debug!("relay re-connecting");
-                self.relay_client.connect().await.context("keepalive")?;
+                relay_client.connect().await.context("keepalive")?;
             }
             tokio::select! {
                 msg = inbox.recv() => {
                     let Some(msg) = msg else {
                         debug!("all clients closed");
                         break;
                     };
-
-                    trace!("tick: inbox: {:?}", msg);
-                    match msg {
-                        ConnectedRelayMessage::GetLastWrite(r) => {
-                            r.send(self.last_write).ok();
-                        }
-                        ConnectedRelayMessage::Ping(r) => {
-                            r.send(self.relay_client.ping().await).ok();
-                        }
-                        ConnectedRelayMessage::GetLocalAddr(r) => {
-                            r.send(self.relay_client.local_addr().await).ok();
-                        }
-                        ConnectedRelayMessage::GetClient(r) => {
-                            self.last_write = Instant::now();
-                            r.send(self.relay_client.clone()).ok();
-                        }
-                        ConnectedRelayMessage::NotePreferred(is_preferred) => {
-                            self.relay_client.note_preferred(is_preferred).await;
-                        }
-                        ConnectedRelayMessage::GetNodeRoute(peer, r) => {
-                            let client = if self.node_present.contains(&peer) {
-                                Some(self.relay_client.clone())
-                            } else {
-                                None
-                            };
-                            r.send(client).ok();
-                        }
-                        ConnectedRelayMessage::Shutdown => {
-                            debug!("shutdown");
-                            break;
-                        }
+                    if self.handle_actor_msg(msg).await {
+                        break;
                     }
                 }
+                _ = &mut relay_send_fut, if matches!(relay_send_fut, Either::Right(_)) => {
+                    relay_send_fut = Either::Left(future::pending());
+                }
+                Some(msg) = self.relay_datagrams_send.recv(), if matches!(relay_send_fut, Either::Left(_)) => {
+                    relay_send_fut = Either::Right(Box::pin(relay_client.send(msg.0, msg.1)));
```

Comment:
why not just spawn these? or sth..this construction is very hard to follow..

Reply:
I really do not want to spawn a task here.  The plan is for `relay_client` to be directly a `AsyncWriter`, probably wrapped in a `Sink` (wrapped in some different client of Client struct).  So this actor's loop has to learn how to send things to the relay server while not deadlocking itself.

My approach here is to start doing this for just sending packets.  The others which send to the relay are `Ping` and `NotePreferred` (and kind of `GetLocalAddr` because it queues up on the same actor inbox right now, but doesn't go all the way to the relay).

If you look at the points at which `Ping` and `GetLocalAddr` are being used: this is only when there network has changed and the `RelayActor` wants to be sure this connection still works.  So at that time there's already a potential hiccup happening.  I totally want to fix those too, but plan to do that by moving those things into the `ConnectedRelayActor` instead.

Reply:
could you use `MaybeFuture` to make this a bit simpler?

Reply:
I made some changes to `MaybeFuture`, agreed that this is nicer.  PTAL

Reply:
much easier to understand (at least for me) now, thanks